README.md
# mogPilot: A fork of jvePilot with a few of my own opinionated changes
This for is only for Mog's Chrysler Pacifica 2021. If it works for you that's great. I still reccomend people run [jvepilot](https://github.com/j-vanetten/openpilot) on which this fork is based.

# Current differences from jvePilot
* Autotether support, enable tether when on road, and disable it when off road
* `steer.checkMinimum` is set to False by defualt (you need to change this to True if you don't have the white panda mod installed)
* Set minimum steer speed to -0.1 so that you don't get warnings at 0mph
* Offset max speed in ui 450 pixels down so its not blocked by my fish eye lens, also dont show max speed until its set.
* Include lkas mitigation from xps https://github.com/commaai/openpilot/pull/21064
* added paho-mqtt as pyextra for future use. op_edit the following variables to make use of.
 + `mqttBroker` mqtt host to connect to
 + `mqttPort` port to connect on defaults to 1883
 + `mqttUser` username to use
 + `mqttPass` password to use

# jvePilot Hybrid OpenPilot/ACC for Chrysler/Jeep 
This fork is only for Chrysler/Jeep vehicles!

[![Buy me a beer!](https://github.com/j-vanetten/openpilot/blob/jvePilot-release/.github/ButMeABeer.png?raw=true)](https://www.buymeacoffee.com/jvePilot)

# Table of contents
- [**Safety Notes**](#safety-notes)
- [jvePilot](#jvepilot)
  * [What is this Fork?](#what-is-this-fork)
  * [Benefits of jvePilot](#benefits-of-jvepilot)
    + [Longitudinal control](#longitudinal-control)
    + [Auto Resume](#auto-resume)
    + [Auto Follow](#auto-follow)
    + [ACC Eco](#acc-eco)
  * [How to use it](#how-to-use-it)
    + [Where to look when setting ACC speed](#where-to-look-when-setting-acc-speed)
- [Install](#install)
  * [Branches](#branches)
- [Customizing](#customizing)
- [Advanced](#advanced-settings)

# **Safety Notes**
* This is my experimental branch, so I'm not responsible for any damage this may cause.
* jvePilot still does not have direct control of the gas and brakes!
  Changing the ACC speed does not always result in the vehicle braking unless the difference in speed is large enough.
  If the speed difference is small, the vehicle just lets off the gas.
* ACC can't go slower that 20mph
* ACC doesn't do a good job at seeing things that are already stopped

---

# jvePilot 
I have a 2018 Grand Cherokee Trailhawk, so I'm only able to confirm features using this vehicle.
* 2017 Gas Chrysler Pacifica: Confirmed by @debugged-hosting

## What is this Fork?
This is my personal OpenPilot fork that includes features that I feel make it a better driving experience for me and possibly others.

## Benefits of jvePilot
* Smother driving in traffic as jvePilot will do a better job at predicting traffic and adjust ACC speed accordingly
* ACC braking by setting the cruse speed lower that the target to help slow the vehicle sooner
* Slow for cars cutting in before ACC does
* Slow in a turn, so you don't have to change the set speed yourself (Speeds are configurable)
* Auto resume after ACC comes to a stop behind vehicle (Can be disabled)
* Auto follow feature to adjust the follow distance based on speed (Speeds are configurable)
* ACC Eco to limit the throttle when accelerating  
* Pressing the gas does not disengage jvePilot (Can be disabled)
* Setting to sync jvePilot display speed to vehicle speedometer 
* Use LKAS button in the dash to disable lane line driving and instead use the new KL driving model. [Read about KL model here](https://blog.comma.ai/end-to-end-lateral-planning).

### Longitudinal control
This fork combines the speed control logic of OpenPilot with the vehicles Adaptive Cruse Control (ACC).
It does this by changing the ACC speed to match the value OpenPilot calculates as the desired speed.
This brings some of OpenPilots longitudinal control to these vehicles.
Including things like slowing while cornering and slowing when it detects cut-ins.
It will also smooth the braking of ACC when driving in traffic.

On FCA vehicles, only the steering is controlled by jvePilot and speed is left up to the ACC of the vehicle.
This fork takes control of the ACC speed setting and adjusts the ACC speed to match the speed jvePilot would be targeting if it actually was able to control the gas and brakes.
It does this by simulating ACC+ and ACC- button presses on the steering wheel to change the ACC speed.
It is limited as ACC only goes down to 20 mph, so it doesn't help as low speeds.

### Auto Resume
ACC will come to a stop behind vehicles, however, if stopped too long, it will either stay stopped until resume is pressed, or simply disengage ACC altogether.  
For the case where ACC simply cancels, the driver has to press and hold the brake to keep the vehicle stopped.
Auto resume makes life easier by resuming ACC when the vehicle in front of you begin to move, or, you let off the brake after coming to a standstill.
While stopped, you can still disengage jvePilot by pressing the Cancel button. 

### Auto Follow
Auto Follow is a way to automate the changing of the stock follow distance setting.
It sets the follow distance to closer at slow speeds and increases it the faster you go.
Setting the follow speed to one/two bars helps with keeping up with cars that take off when stopped or at slow speeds.
The faster you go, the more distance you want, so you can have more confidence in ACC being able to stop in case it needs to.

The current enabled state of Auto Follow is displayed on the bottom of the jvePilot display.
Pressing Follow + or - will disable Auto Follow giving you full control to set the follow distance. 
To re-enable Auto Follow, hold either Follow + or - for half a second or tap the button on the display. 
 
### ACC Eco
When enabled, jvePilot will limit how far ahead the ACC setting is above the current speed.  
This prevents the vehicle from using an aggressive throttle to get up to speed saving on gas/battery.

The ACC Eco button is located in the lower right corner of the display.  
Tapping the button cycles between off, level 1, and level 2 eco settings.
Level 2 provides the slowest acceleration and is selected when both leaves are green.    
Level 1 should provide a balance is selected when only one leaf is green.
If you feel these settings are not right for you or your vehicle, see the [ACC Eco](#acc-eco) setting to adjust them. 
Much like your vehicles eco/sport modes, the current setting is persisted between drives.

## How to use it 
When using this branch, you will be setting the max ACC speed on the jvePilot display instead of the one in the dashboard.
jvePilot will then set the ACC setting in the dashboard to the targeted speed, but never exceeding the max speed set on the jvePilot display.
A quick press of the ACC+ and ACC- buttons will change this speed by 5 mph on the jvePilot display, while a long deliberate press (about a 1/2 second press) changes it by 1 mph.
DO NOT hold the ACC+ or ACC- buttons for longer that a 1 second. Either make quick or long deliberate presses only.

### Where to look when setting ACC speed
Do not look at the dashboard when setting your ACC max speed.
Instead, only look at the one on the jvePilot display.
The reason you need to look at jvePilot is because jvePilot will be changing the one in the dashboard.
It will be adjusting it as needed, never raising it above the one set on the jvePilot display.

**ONLY look at the MAX speed on jvePilot when setting the ACC speed instead of the dashboard!**

---

# Install
The easiest way to install jvePilot is to factory reset and use this Custom Software URL: `https://bit.ly/jvePilot-release`

## Branches
`/jvePilot-release` - The latest release.  Will contain the latest version I feel is ready for daily use. Custom Software URL: `https://bit.ly/jvePilot-release`
`/jvePilot-beta` - Sometimes I have people wanting to beta test jvePilot's new features.  Mostly stable, but still can be buggy. Custom Software URL: `https://bit.ly/jvePilot-beta` 
`/feature/*` - These branches are where I'm working on new features.  These are never safe to run as they change all the time.

---
# Customizing
Customizing features and parameters can be done on the UI display.  
Click the gear icon and then select jvePilot from the sidebar.  
Clicking on the text of feature of will show more information about it and allow customization of the feature.
Note that some settings do require a vehicle restart, and some don't 
The settings that don't require a restart may take upward of 5 seconds to take effect after changing them. 


## Slow in Curves
jvePilot will slow in curves so that you don't have to.
* Default: On
* Vehicle Restart Required: No
### Speed Ratio 
Use this to tune the speed in curves to you liking.
Setting this to 1.2 will cause jvePilot to drive 20% faster in turns than if it was set to the default of 1.0
* Default: 1.0
* Units: Ratio
* Vehicle Restart Required: No
* Min/Max values (0.1, 2)
### Drop off
Adjusts how much the speed drops as the curve increases.
Decrease this value to lessen the amount of drop off as the curve increases.
Changing this value will likely require adjusting the Speed Ratio to compensate.
* Default: 2.0
* Vehicle Restart Required: No
* Min/Max values (1.0, 3.0)

## Reverse ACC +/- Speeds
Reverse the stock ACC +/- button's 1mph on short press and 5mph on long press.  Turn off to return to stock style.
* Default: On
* Vehicle Restart Required: Yes

## Auto Resume
This feature allows jvePilot to auto resume from an ACC stop.
* Default: On
* Vehicle Restart Required: Yes

## Disable on Gas
When enabled, jvePilot will disengage when you press the gas
* Default: Off
* Vehicle Restart Required: Yes

## Auto Follow
If you don't want auto follow enabled on every start, turn this off.
### 1-2 Bar Change Over
When your speed (in MPH) is below this setting, Auto Follow will set the follow setting to one bar.  
When you reach this speed (in MPH), Auto Follow will set the follow setting to two bars.
* Default: 15
* Units: MPH
* Vehicle Restart Required: No
* Min/Max values (0, 300)
### 2-3 Bar Change Over
When your speed (in MPH) is below this setting, Auto Follow will set the follow setting to two bars.
When you reach this speed (in MPH), Auto Follow will set the follow setting to three bars.
* Default: 30
* Units: MPH
* Vehicle Restart Required: No
* Min/Max values (0, 300)
### 3-4 Bar Change Over
When your speed (in MPH) is below this setting, Auto Follow will set the follow setting to three bars.
When you reach this speed (in MPH), Auto Follow will set the follow setting to four bars.
* Default: 65
* Units: MPH
* Vehicle Restart Required: No
* Min/Max values (0, 300)

## ACC Eco
ACC Eco limits acceleration by keep the ACC cruise speed closer to your current speed.
These setting are how far ahead, in MPH, of your current speed ACC will be set.  
The higher the number, the more aggressive ACC will be when accelerating.
### Keep ahead at ACC Eco level 1
Use this setting to adjust ACC Eco level 1 (one green leaf) for a balance of speed and eco-ness  
* Default: 7
* Units: MPH
* Vehicle Restart Required: No
* Min/Max values 1, 100
### Keep ahead at ACC Eco level 2
Use this setting to adjust ACC Eco level 2 (two green leaves) for maximum eco-ness
* Default: 5
* Units: MPH
* Vehicle Restart Required: No
* Min/Max values 1, 100

## jvePilot Control Settings
### Device Offset
Compensate for mounting your device off center in the windshield.
If you mounted your device off center, use this setting to compensate.
Use positive values if your device is to the left of center and negative if it's to the right.
* Default: 0
* Units: Meters
* Vehicle Restart Required: No
* Min/Max values -1, 1
### Speed Adjust Ratio
Adjust speed displayed by jvePilot to match the real world.

RELEASES.md
Version jvePilot-0.8.14(v3.10.0)
========================
Visit http://bit.ly/jvereadme for more information
* New radar code from xps's fork
* Restore 9mph min steer on some models 
* Update to 0.8.13
  * New driving model
    * Bigger model, using both of comma three's road-facing cameras
    * Better at cut-in detection and tight turns
  * New driver monitoring model
    * Tweaked network structure to improve output resolution for DSP
    * Fixed bug in quantization aware training to reduce quantizing errors
    * Resulted in 7x less MSE and no more random biases at runtime
  * Added toggle to disable disengaging on the accelerator pedal
  * comma body support
  * Audi RS3 support thanks to jyoung8607!
  * Hyundai Ioniq Plug-in Hybrid 2019 support thanks to sunnyhaibin!
  * Hyundai Tucson Diesel 2019 support thanks to sunnyhaibin!
  * Toyota Alphard Hybrid 2021 support
  * Toyota Avalon Hybrid 2022 support
  * Toyota RAV4 2022 support
  * Toyota RAV4 Hybrid 2022 support

Version jvePilot-0.8.13(v3.9.0)
========================
* Reverse lateral angle of radar on Pacifica's
* Update to 0.8.13
  * Improved driver monitoring
    * Retuned driver pose learner for relaxed driving positions
    * Added reliance on driving model to be more scene adaptive
    * Matched strictness between comma two and comma three
  * Improved performance in turns by compensating for the road bank angle
  * Improved camera focus on the comma two
  * AGNOS 4
    * ADB support
    * improved cell auto configuration
  * NEOS 19
    * package updates
    * stability improvements
  * Subaru ECU firmware fingerprinting thanks to martinl!
  * Hyundai Santa Fe Plug-in Hybrid 2022 support thanks to sunnyhaibin!
  * Mazda CX-5 2022 support thanks to Jafaral!
  * Subaru Impreza 2020 support thanks to martinl!
  * Toyota Avalon 2022 support thanks to sshane!
  * Toyota Prius v 2017 support thanks to CT921!
  * Volkswagen Caravelle 2020 support thanks to jyoung8607!

Version jvePilot-0.8.12(v3.8.0)
========================
* Turn Auto Follow into a small icon above ACC Eco button
  * Use buttons to toggle on/off, not screen
* Fixed radar.  Should get less collision alert
* Update to 0.8.12
  * New driving model
    * Improved behavior around exits
    * Better pose accuracy at high speeds, allowing max speed of 90mph
    * Fully incorporated comma three data into all parts of training stack
  * Improved follow distance
  * Better longitudinal policy, especially in low speed traffic
  * New alert sounds
  * AGNOS 3
    * Display burn in mitigation
    * Improved audio amplifier configuration
    * System reliability improvements
    * Update Python to 3.8.10
  * Raw logs upload moved to connect.comma.ai
  * Fixed HUD alerts on newer Honda Bosch thanks to csouers!
  * Audi Q3 2020-21 support thanks to jyoung8607!
  * Lexus RC 2020 support thanks to ErichMoraga!
* Update to 0.8.11
  * Support for CAN FD on the red panda
  * Support for an external panda on the comma three
  * Navigation: Show more detailed instructions when approaching maneuver
  * Fixed occasional steering faults on GM cars thanks to jyoung8607!
  * Nissan ECU firmware fingerprinting thanks to robin-reckmann, martinl, and razem-io!
  * Cadillac Escalade ESV 2016 support thanks to Gibby!
  * Genesis G70 2020 support thanks to tecandrew!
  * Hyundai Santa Fe Hybrid 2022 support thanks to sunnyhaibin!
  * Mazda CX-9 2021 support thanks to Jacar!
  * Volkswagen Polo 2020 support thanks to jyoung8607!
  * Volkswagen T-Roc 2021 support thanks to jyoung8607!

Version jvePilot-0.8.10(v3.7)
========================
* Crude replacement to continue to support stock ACC follow distance profiles
  * OP removed the use of radar to track lead car.  This broke how I was supporting distance profiles.
  * So, to fix, always do +3 mph over OP targeted speed to force stock ACC to keep the distance
* Speed turns green when gas applied, red when braking
* Fix for White Panda mod on 2017/2018 models
* Fix bug where left turns were not slowing
* Update to 0.8.10
  * New driving model
    * Trained on one million minutes!!!
    * Fixed lead training making lead predictions significantly more accurate
    * Fixed several localizer dataset bugs and loss function bugs, overall improved accuracy
  * New driver monitoring model
    * Trained on latest data from both comma two and comma three
    * Increased model field of view by 40% on comma three
    * Improved model stability on masked users
    * Improved pose prediction with reworked ground-truth stack
  * Lateral and longitudinal planning MPCs now in ACADOS
  * Combined longitudinal MPCs
    * All longitudinal planning now happens in a single MPC system
    * Fixed instability in MPC problem to prevent sporadic CPU usage
  * AGNOS 2: minor stability improvements and builder repo open sourced
  * tools: new and improved replay thanks to deanlee!
  * Moved community-supported cars outside of the Community Features toggle
  * Improved FW fingerprinting reliability for Hyundai/Kia/Genesis
  * Added prerequisites for longitudinal control on Hyundai/Kia/Genesis and Honda Bosch
  * Audi S3 2015 support thanks to jyoung8607!
  * Honda Freed 2020 support thanks to belm0!
  * Hyundai Ioniq Hybrid 2020-2022 support thanks to sunnyhaibin!
  * Hyundai Santa Fe 2022 support thanks to sunnyhaibin!
  * Kia K5 2021 support thanks to sunnyhaibin!
  * Škoda Kamiq 2021 support thanks to jyoung8607!
  * Škoda Karoq 2019 support thanks to jyoung8607!
  * Volkswagen Arteon 2021 support thanks to jyoung8607!
  * Volkswagen California 2021 support thanks to jyoung8607!
  * Volkswagen Taos 2022 support thanks to jyoung8607!

* Version jvePilot-0.8.9(v3.6.1)
========================
* Add option to alert when steering is lost do to low speeds
* Fix for White Panda mod lkas fault when jvePilot is not engaged.

selfdrive/ui/paint.cc
#include "selfdrive/ui/paint.h"

#include <cassert>

#ifdef __APPLE__
#include <OpenGL/gl3.h>
#define NANOVG_GL3_IMPLEMENTATION
#define nvgCreate nvgCreateGL3
#else
#include <GLES3/gl3.h>
#define NANOVG_GLES3_IMPLEMENTATION
#define nvgCreate nvgCreateGLES3
#endif

#define NANOVG_GLES3_IMPLEMENTATION
#include <nanovg_gl.h>
#include <nanovg_gl_utils.h>

#include "selfdrive/common/util.h"
#include "selfdrive/hardware/hw.h"
#include "selfdrive/ui/ui.h"

static void ui_draw_text(const UIState *s, float x, float y, const char *string, float size, NVGcolor color, const char *font_name) {
  nvgFontFace(s->vg, font_name);
  nvgFontSize(s->vg, size);
  nvgFillColor(s->vg, color);
  nvgText(s->vg, x, y, string, NULL);
}

static void draw_chevron(UIState *s, float x, float y, float sz, NVGcolor fillColor, NVGcolor glowColor) {
  // glow
  float g_xo = sz/5;
  float g_yo = sz/10;
  nvgBeginPath(s->vg);
  nvgMoveTo(s->vg, x+(sz*1.35)+g_xo, y+sz+g_yo);
  nvgLineTo(s->vg, x, y-g_xo);
  nvgLineTo(s->vg, x-(sz*1.35)-g_xo, y+sz+g_yo);
  nvgClosePath(s->vg);
  nvgFillColor(s->vg, glowColor);
  nvgFill(s->vg);

  // chevron
  nvgBeginPath(s->vg);
  nvgMoveTo(s->vg, x+(sz*1.25), y+sz);
  nvgLineTo(s->vg, x, y);
  nvgLineTo(s->vg, x-(sz*1.25), y+sz);
  nvgClosePath(s->vg);
  nvgFillColor(s->vg, fillColor);
  nvgFill(s->vg);
}

static void ui_draw_circle_image(const UIState *s, int center_x, int center_y, int radius, const char *image, NVGcolor color, float img_alpha) {
  nvgBeginPath(s->vg);
  nvgCircle(s->vg, center_x, center_y, radius);
  nvgFillColor(s->vg, color);
  nvgFill(s->vg);
  const int img_size = radius * 1.5;
  ui_draw_image(s, {center_x - (img_size / 2), center_y - (img_size / 2), img_size, img_size}, image, img_alpha);
}

static void ui_draw_circle_image(const UIState *s, int center_x, int center_y, int radius, const char *image, bool active) {
  float bg_alpha = active ? 0.3f : 0.1f;
  float img_alpha = active ? 1.0f : 0.15f;
  ui_draw_circle_image(s, center_x, center_y, radius, image, nvgRGBA(0, 0, 0, (255 * bg_alpha)), img_alpha);
}

static void draw_lead(UIState *s, const cereal::ModelDataV2::LeadDataV3::Reader &lead_data, const vertex_data &vd) {
  // Draw lead car indicator
  auto [x, y] = vd;

  float fillAlpha = 0;
  float speedBuff = 10.;
  float leadBuff = 40.;
  float d_rel = lead_data.getX()[0];
  float v_rel = lead_data.getV()[0];
  if (d_rel < leadBuff) {
    fillAlpha = 255*(1.0-(d_rel/leadBuff));
    if (v_rel < 0) {
      fillAlpha += 255*(-1*(v_rel/speedBuff));
    }
    fillAlpha = (int)(fmin(fillAlpha, 255));
  }

  float sz = std::clamp((25 * 30) / (d_rel / 3 + 30), 15.0f, 30.0f) * 2.35;
  x = std::clamp(x, 0.f, s->fb_w - sz / 2);
  y = std::fmin(s->fb_h - sz * .6, y);
  draw_chevron(s, x, y, sz, nvgRGBA(201, 34, 49, fillAlpha), COLOR_YELLOW);
}

static void ui_draw_line(UIState *s, const line_vertices_data &vd, NVGcolor *color, NVGpaint *paint) {
  if (vd.cnt == 0) return;

  const vertex_data *v = &vd.v[0];
  nvgBeginPath(s->vg);
  nvgMoveTo(s->vg, v[0].x, v[0].y);
  for (int i = 1; i < vd.cnt; i++) {
    nvgLineTo(s->vg, v[i].x, v[i].y);
  }
  nvgClosePath(s->vg);
  if (color) {
    nvgFillColor(s->vg, *color);
  } else if (paint) {
    nvgFillPaint(s->vg, *paint);
  }
  nvgFill(s->vg);
}

static void ui_draw_vision_lane_lines(UIState *s) {
  const UIScene &scene = s->scene;
  NVGpaint track_bg;
  if (!scene.end_to_end) {
    // paint lanelines
    for (int i = 0; i < std::size(scene.lane_line_vertices); i++) {
      NVGcolor color = nvgRGBAf(1.0, 1.0, 1.0, scene.lane_line_probs[i]);
      ui_draw_line(s, scene.lane_line_vertices[i], &color, nullptr);
    }

    // paint road edges
    for (int i = 0; i < std::size(scene.road_edge_vertices); i++) {
      NVGcolor color = nvgRGBAf(1.0, 0.0, 0.0, std::clamp<float>(1.0 - scene.road_edge_stds[i], 0.0, 1.0));
      ui_draw_line(s, scene.road_edge_vertices[i], &color, nullptr);
    }
    track_bg = nvgLinearGradient(s->vg, s->fb_w, s->fb_h, s->fb_w, s->fb_h * .4,
                                          COLOR_WHITE, COLOR_WHITE_ALPHA(0));
  } else {
    track_bg = nvgLinearGradient(s->vg, s->fb_w, s->fb_h, s->fb_w, s->fb_h * .4,
                                          COLOR_RED, COLOR_RED_ALPHA(0));
  }
  // paint path
  ui_draw_line(s, scene.track_vertices, nullptr, &track_bg);
}

// Draw all world space objects.
static void ui_draw_world(UIState *s) {
  nvgScissor(s->vg, 0, 0, s->fb_w, s->fb_h);

  // Draw lane edges and vision/mpc tracks
  ui_draw_vision_lane_lines(s);

  // Draw lead indicators if openpilot is handling longitudinal
  if (s->scene.longitudinal_control) {
    auto lead_one = (*s->sm)["modelV2"].getModelV2().getLeadsV3()[0];
    auto lead_two = (*s->sm)["modelV2"].getModelV2().getLeadsV3()[1];
    if (lead_one.getProb() > .5) {
      draw_lead(s, lead_one, s->scene.lead_vertices[0]);
    }
    if (lead_two.getProb() > .5 && (std::abs(lead_one.getX()[0] - lead_two.getX()[0]) > 3.0)) {
      draw_lead(s, lead_two, s->scene.lead_vertices[1]);
    }
  }
  nvgResetScissor(s->vg);
}

static void ui_draw_vision_maxspeed(UIState *s) {
  const int SET_SPEED_NA = 255;
  const int off_x = 0;
  const int off_y = 450;

  float maxspeed = (*s->sm)["controlsState"].getControlsState().getVCruise();
  const bool is_cruise_set = maxspeed != 0 && maxspeed != SET_SPEED_NA;
  if (is_cruise_set && !s->scene.is_metric) { maxspeed *= 0.6225; }

  const Rect rect = {(bdr_s * 2) + off_x, int(bdr_s * 1.5) + off_y, 184, 202};
  if(!is_cruise_set)
    return;

  ui_fill_rect(s->vg, rect, COLOR_BLACK_ALPHA(100), 30.);
  ui_draw_rect(s->vg, rect, COLOR_WHITE_ALPHA(100), 10, 20.);

  nvgTextAlign(s->vg, NVG_ALIGN_CENTER | NVG_ALIGN_BASELINE);
  ui_draw_text(s, rect.centerX(), 118 + off_y, "MAX", 26 * 2.5, COLOR_WHITE_ALPHA(is_cruise_set ? 200 : 100), "sans-regular");
  if (is_cruise_set) {
    const std::string maxspeed_str = std::to_string((int)std::nearbyint(maxspeed));
    ui_draw_text(s, rect.centerX(), 212 + off_y, maxspeed_str.c_str(), 48 * 2.5, COLOR_WHITE, "sans-bold");
  } else {
    ui_draw_text(s, rect.centerX(), 212 + off_y, "N/A", 42 * 2.5, COLOR_WHITE_ALPHA(100), "sans-semibold");
  }
}

static void ui_draw_vision_speed(UIState *s) {
  const float speed = std::max(0.0, (*s->sm)["carState"].getCarState().getVEgo() * (s->scene.is_metric ? 3.6 : 2.2369363));
  const float pedalPressedAmount = (*s->sm)["carState"].getCarState().getJvePilotCarState().getPedalPressedAmount();
  const std::string speed_str = std::to_string((int)std::nearbyint(speed));
  nvgTextAlign(s->vg, NVG_ALIGN_CENTER | NVG_ALIGN_BASELINE);
  ui_draw_text(s, s->fb_w/2, 210, speed_str.c_str(), 96 * 2.5, COLOR_WHITE, "sans-bold");
  if (pedalPressedAmount < 0) {
    ui_draw_text(s, s->fb_w/2, 210, speed_str.c_str(), 96 * 2.5, COLOR_SPEED_BRAKE_ALPHA(pedalPressedAmount * -256), "sans-bold");
  } else if (pedalPressedAmount > 0) {
    ui_draw_text(s, s->fb_w/2, 210, speed_str.c_str(), 96 * 2.5, COLOR_SPEED_GAS_ALPHA(pedalPressedAmount * 256), "sans-bold");
  }
  ui_draw_text(s, s->fb_w/2, 290, s->scene.is_metric ? "km/h" : "mph", 36 * 2.5, COLOR_WHITE_ALPHA(200), "sans-regular");
}

static void ui_draw_vision_event(UIState *s) {
  if (s->scene.engageable) {
    // draw steering wheel
    const int radius = 96;
    const int center_x = s->fb_w - radius - bdr_s * 2;
    const int center_y = radius  + (bdr_s * 1.5);
    const QColor &color = bg_colors[s->status];
    NVGcolor nvg_color = nvgRGBA(color.red(), color.green(), color.blue(), color.alpha());
    ui_draw_circle_image(s, center_x, center_y, radius, "wheel", nvg_color, 1.0f);
  }
}

static void ui_draw_vision_face(UIState *s) {
  const int radius = 96;
  const int center_x = radius + (bdr_s * 2);
  const int center_y = s->fb_h - footer_h / 2;
  ui_draw_circle_image(s, center_x, center_y, radius, "driver_face", s->scene.dm_active);
}

static void ui_draw_button(UIState *s, Rect btn, const char *text, bool enabled) {
  int btn_x = btn.x;
  int btn_y = btn.y;
  int btn_w = btn.w;
  int btn_h = btn.h;
  int center_x = btn_x + (btn_w / 2);
  int center_y = btn_y + (btn_h / 2);

  nvgBeginPath(s->vg);
  nvgRoundedRect(s->vg, btn_x, btn_y, btn_w, btn_h, 25);
  if (enabled) {  // change outline color based on status of button
    nvgStrokeColor(s->vg, nvgRGBA(55, 184, 104, 255));
  } else {
    nvgStrokeColor(s->vg, nvgRGBA(200, 200, 200, 200));
  }
  nvgStrokeWidth(s->vg, 12);
  nvgStroke(s->vg);

  nvgBeginPath(s->vg);  // dark background for readability
  nvgRoundedRect(s->vg, btn_x, btn_y, btn_w, btn_h, 25);
  if (enabled) {  // change background color based on status of button
    nvgFillColor(s->vg, nvgRGBA(55, 184, 104, 75));
  } else {
    nvgFillColor(s->vg, nvgRGBA(0, 0, 0, 75));
  }
  nvgFill(s->vg);

  nvgFillColor(s->vg, nvgRGBA(255, 255, 255, 255));
  nvgFontSize(s->vg, 60);
  nvgText(s->vg, center_x, center_y + btn_h / 8, text, NULL);
}

static void ui_draw_jve_pilot_buttons(UIState *s) {
  if (s->scene.autoFollowEnabled != -1) {
    bool autoFollowEnabled = s->scene.autoFollowEnabled == 1;

    const char *message;
    if (s->scene.autoFollow_btn.w == autoFollowSmallWidth_btn) {
      message = autoFollowEnabled ? "AF ON" : "AF OFF";
    } else {
      message = autoFollowEnabled ? "Auto Follow" : "Auto Follow OFF";
    }

    ui_draw_button(s, s->scene.autoFollow_btn, message, autoFollowEnabled);
  }
  if (s->scene.accEco != -1) {
    const char *img = s->scene.accEco == 1 ? "acc_eco_1" : s->scene.accEco == 2 ? "acc_eco_2" : "acc_eco_off";
    ui_draw_image(s, s->scene.accEco_img, img, s->scene.accEco != 0 ? 1.0f : 0.7f);
  }
}

static void ui_draw_vision_header(UIState *s) {
  NVGpaint gradient = nvgLinearGradient(s->vg, 0, header_h - (header_h / 2.5), 0, header_h,
                                        nvgRGBAf(0, 0, 0, 0.45), nvgRGBAf(0, 0, 0, 0));
  ui_fill_rect(s->vg, {0, 0, s->fb_w , header_h}, gradient);
  ui_draw_vision_maxspeed(s);
  ui_draw_vision_speed(s);
  ui_draw_vision_event(s);

  ui_draw_jve_pilot_buttons(s);
}

static void ui_draw_vision(UIState *s) {
  const UIScene *scene = &s->scene;
  // Draw augmented elements
  if (scene->world_objects_visible) {
    ui_draw_world(s);
  }
  // Set Speed, Current Speed, Status/Events
  ui_draw_vision_header(s);
  if ((*s->sm)["controlsState"].getControlsState().getAlertSize() == cereal::ControlsState::AlertSize::NONE) {
    ui_draw_vision_face(s);
  }
}

void ui_draw(UIState *s, int w, int h) {
  // Update intrinsics matrix after possible wide camera toggle change
  if (s->fb_w != w || s->fb_h != h) {
    ui_resize(s, w, h);
  }
  glEnable(GL_BLEND);
  glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
  nvgBeginFrame(s->vg, s->fb_w, s->fb_h, 1.0f);
  ui_draw_vision(s);
  nvgEndFrame(s->vg);
  glDisable(GL_BLEND);
}

void ui_draw_image(const UIState *s, const Rect &r, const char *name, float alpha) {
  nvgBeginPath(s->vg);
  NVGpaint imgPaint = nvgImagePattern(s->vg, r.x, r.y, r.w, r.h, 0, s->images.at(name), alpha);
  nvgRect(s->vg, r.x, r.y, r.w, r.h);
  nvgFillPaint(s->vg, imgPaint);
  nvgFill(s->vg);
}

void ui_draw_rect(NVGcontext *vg, const Rect &r, NVGcolor color, int width, float radius) {
  nvgBeginPath(vg);
  radius > 0 ? nvgRoundedRect(vg, r.x, r.y, r.w, r.h, radius) : nvgRect(vg, r.x, r.y, r.w, r.h);
  nvgStrokeColor(vg, color);
  nvgStrokeWidth(vg, width);
  nvgStroke(vg);
}

static inline void fill_rect(NVGcontext *vg, const Rect &r, const NVGcolor *color, const NVGpaint *paint, float radius) {
  nvgBeginPath(vg);
  radius > 0 ? nvgRoundedRect(vg, r.x, r.y, r.w, r.h, radius) : nvgRect(vg, r.x, r.y, r.w, r.h);
  if (color) nvgFillColor(vg, *color);
  if (paint) nvgFillPaint(vg, *paint);
  nvgFill(vg);
}
void ui_fill_rect(NVGcontext *vg, const Rect &r, const NVGcolor &color, float radius) {
  fill_rect(vg, r, &color, nullptr, radius);
}
void ui_fill_rect(NVGcontext *vg, const Rect &r, const NVGpaint &paint, float radius) {
  fill_rect(vg, r, nullptr, &paint, radius);
}

void ui_nvg_init(UIState *s) {
  // on EON, we enable MSAA
  s->vg = Hardware::EON() ? nvgCreate(0) : nvgCreate(NVG_ANTIALIAS | NVG_STENCIL_STROKES | NVG_DEBUG);
  assert(s->vg);

  // init fonts
  std::pair<const char *, const char *> fonts[] = {
      {"sans-regular", "../assets/fonts/opensans_regular.ttf"},
      {"sans-semibold", "../assets/fonts/opensans_semibold.ttf"},
      {"sans-bold", "../assets/fonts/opensans_bold.ttf"},
  };
  for (auto [name, file] : fonts) {
    int font_id = nvgCreateFont(s->vg, name, file);
    assert(font_id >= 0);
  }

  // init images
  std::vector<std::pair<const char *, const char *>> images = {
    {"acc_eco_1", "../assets/jvepilot/img_acc_eco_1.png"},
    {"acc_eco_2", "../assets/jvepilot/img_acc_eco_2.png"},
    {"acc_eco_off", "../assets/jvepilot/img_acc_eco_off.png"},
    {"wheel", "../assets/img_chffr_wheel.png"},
    {"driver_face", "../assets/img_driver_face.png"},
  };
  for (auto [name, file] : images) {
    s->images[name] = nvgCreateImage(s->vg, file, 1);
    assert(s->images[name] != 0);
  }
}

void ui_resize(UIState *s, int width, int height) {
  s->fb_w = width;
  s->fb_h = height;

  int autoFollowWidth_btn = width < 1200 ? autoFollowSmallWidth_btn : autoFollowLargeWidth_btn;
  s->scene.autoFollow_btn = {width/2 - autoFollowWidth_btn/2, height - v_edge_padding - 130/2, autoFollowWidth_btn, 130};
  s->scene.accEco_img = {width - h_edge_padding - 233, height - v_edge_padding/2 - 233, 233, 233};

  auto intrinsic_matrix = s->wide_camera ? ecam_intrinsic_matrix : fcam_intrinsic_matrix;
  float zoom = ZOOM / intrinsic_matrix.v[0];
  if (s->wide_camera) {
    zoom *= 0.5;
  }

  // Apply transformation such that video pixel coordinates match video
  // 1) Put (0, 0) in the middle of the video
  nvgTranslate(s->vg, width / 2, height / 2 + y_offset);
  // 2) Apply same scaling as video
  nvgScale(s->vg, zoom, zoom);
  // 3) Put (0, 0) in top left corner of video
  nvgTranslate(s->vg, -intrinsic_matrix.v[2], -intrinsic_matrix.v[5]);

  nvgCurrentTransform(s->vg, s->car_space_transform);
  nvgResetTransform(s->vg);
}
